presto-hive/src/main/java/com/facebook/presto/hive/HiveStagingFileCommitter.java

```diff
@@ -0,0 +1,86 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.hive.HdfsEnvironment.HdfsContext;
+import com.facebook.presto.hive.PartitionUpdate.StagingFileInfo;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.SchemaTableName;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_FILESYSTEM_ERROR;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HiveStagingFileCommitter
+        implements StagingFileCommitter
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final Executor renameExecutor;
+
+    public HiveStagingFileCommitter(
+            HdfsEnvironment hdfsEnvironment,
+            Executor renameExecutor)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.renameExecutor = requireNonNull(renameExecutor, "renameExecutor is null");
+    }
+
+    @Override
+    public void commitFiles(ConnectorSession session, SchemaTableName tableName, List<PartitionUpdate> fragments)
+    {
+        List<Future<?>> fileRenameFutures = new ArrayList<>();
+        for (PartitionUpdate partitionUpdate : fragments) {
+            Path path = partitionUpdate.getWritePath();
+            HdfsContext context = new HdfsContext(session, tableName.getSchemaName(), tableName.getTableName());
+            FileSystem fileSystem;
+            try {
+                fileSystem = hdfsEnvironment.getFileSystem(context, path);
+            }
+            catch (IOException e) {
+                throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error moving data files to final location. Error listing directory %s", path), e);
+            }
+            for (StagingFileInfo stagingFileInfo : partitionUpdate.getStagingFileInfos()) {
+                fileRenameFutures.add(CompletableFuture.runAsync(() -> renameFile(fileSystem, path, stagingFileInfo), renameExecutor));
+            }
+        }
+        for (Future<?> fileRenameFuture : fileRenameFutures) {
```

Comment:
This loop won't fail fast. We should fail fast here. Use `Futures.allAsList` from guava. Don't forget to interruptably cancel the list future in case of a failure.

Reply:
I think `ListenableFuture` are generally preferred to be used in Presto codebase. You can then use methods like `whenAllSucceed` instead of a for loop :)

Reply:
@arhimondr : Do you have any methods that allows fail fast in mind ? :)

Reply:
Yeah, we can use the `whenAllSucceed` as well (if fails when any of the futures fail). I just thought of the `allAsList` because the `whenAllSucceed` is a pretty new one i didn't know about.

Reply:
In `SemiTransactionalHiveMetastore#asyncRename`, it puts `CompletableFuture` into a list and then uses similar for-loop to wait for them `SemiTransactionalHiveMetastore#waitForAsyncRenames`. 

But looking at `CompletableFuture`'s documentation, I don't think it could cancel the thread because `interrupt` is not used to control processing.

Reply:
Still i think it is better to fail fast, and at least try to cancel other renames if any fails. We can change the `SemiTransactionalHiveMetastore#asyncRename` in a separate commit as well.

>  I'll first change to CompletableFuture, and if you have strong opinion on using ListenableFuture, we could rework from the executor.

AFAIK we are trying to avoid CompletableFuture. The `ListenableFuture` has much better interface. `CompletableFuture` is terrible.

Reply:
> I don't think it could cancel the thread because interrupt is not used to control processing

If you used `whenAllSucceed`, the cancellation on the returned future propagates to the undelying futures

Reply:
We prefer `ListenableFuture` instead of `CompletableFuture` in our codebase for various reasons, please see https://github.com/prestodb/presto/pull/7540, so I think you should update to use `ListenableFuture`. 

Also please make sure to cancel all futures if any of them fails.